### DISCUSSION TITLE
Begin executing task without creating temp task variable and await its result when needed

### DISCUSSION THREAD

**Discussion by Rekkonnect** [dotnet/csharplang#4912] [Language Ideas]
Currently, we've got the following solution to simultaneously perform operations while a task is being executed:

```csharp
var task = SomeTask();
// other operations
var result = await task;
```

I believe that this could see some syntax sugar improvements, like so:

```csharp
var result = async await SomeTask();
// other operations
await get result;
```

Which could work for more individual variables too:

```csharp
var result1 = async await SomeTask(1);
var result2 = async await SomeTask(2);
await get result1, result2;
```

The above example currently is available like so:

```csharp
var task1 = SomeTask(1);
var task2 = SomeTask(2);
Task.WaitAll(task1, task2);
var result1 = task1;
var result2 = task2;
```

# Edge cases

Variables assigned with `async await` are marked as `async get`, which enables the following features:
- Using an `async get` variable anywhere before using `await get` on it will yield a warning about potentially using an unassigned value.
  - In that case, the compiler will use the backed task instance's `Result` property over the variable's value.
- Functions can accept `async get` parameters, which won't emit the above warning when such a variable is passed, and enabling usage of `await get` on the passed argument.

## Examples

Assume a function `T Transform<T>(T)` that transforms a value in some way and returns the transformed result.

```csharp
var result = async await SomeTask();
// operations
var transformed = Transform(result);
// Warning about using result before using await get on it
```

Now, if that same `Transform` function's first parameter were to accept `async get` variables, the warning will not be emitted, and would allow the function to use `await get` on the parameter.

```csharp
T Transform<T>(async get T value)
{
    await get value;
    // transformation process and returning the result
}
```

You can also pass normal variables in `async get` parameters, in which case a backing function without the `async get` modifier in the respective parameter is created. This also removes all `await get` statements on the respective parameters. For example:

```csharp
// assume
void Function<T>(async get T value)
{
    // operations A
    await get value;
    // operations B
}

// Valid:
Function(1);

// Compiler also generates and uses:
void Function<T>(T value)
{
    // operations A
    // operations B
}
```

---

The proposed syntax is
- `async await` to indicate awaiting the result in a different thread, while simultaneously not blocking execution of the function
- `await get` to block the current thread for the remaining of the execution of the task, if any, ensuring that the result will be assigned to the respective variable
- `async get` to denote that `await get` can be applied on an identifier.

The exact keywords that are to be used are subject to change depending on preferences and opinions.

**Comment by jnm2** [comment]
If you add https://www.nuget.org/packages/TaskTupleAwaiter/ to your project, you can do this:

```cs
var (result1, result2) = await (SomeTask(1), SomeTask(2));
```

You can even use `.ConfigureAwait` on the tuple if you like.

Moving work to a new thread should be done explicitly with `Task.Run` and should not be something the compiler has syntax for. If you're afraid SomeTask(1) will not go asynchronous soon enough and will block SomeTask(2) from starting for a significant amount of time, I would do this:

```cs
var (result1, result2) = await (
    Task.Run(() => SomeTask(1)),
    SomeTask(2));
```

**Reply by Rekkonnect** [reply]
Kudos for introducing me to this library.

However, this does not cover all the cases for the feature. I did also show how there can be other use cases. `Task.Run` is great for performing operations whose results you do not want to await, or don't even return any results in the first place. When you do want their results, but can execute them simultaneously because of context requiring several operations be performed in the background, this is where it can come handy.

Moreover, TupleAwaiter can only be so useful in simple cases where expressions are not that large, and no more than a few tasks are performed. It's great for short expressions and 2 or 3 tasks, but any more will cause worse code quality than it generally would help offering, dragging you back to the starting point.

Again, it is not something that necessarily violates any principles, and can come in handy in select cases where `await` seems to be too commonplace. It helps reduce the clutter of common patterns that are correctly used in those cases.

**Reply by jnm2** [reply]
Can you show me an example then? I'm not seeing one where Task.Run and TaskTupleAwaiter wouldn't work well. 
Besides that, I have reservations about there being syntax to move work to a new thread and about `await get` blocking a thread.

`Task.Run` is not primarily great for performing operations whose results you do not want to await. `Task.Run` comes in handy whether or not you want the results. Even if you don't want the results, it's bad practice not to either await it or hook up immediate error handling, for several reasons.

I routinely use TaskTupleAwaiter with anywhere from 2 to 10 tuple elements. I don't think it's true that it causes worse code quality or drags you back to the starting point.

**Reply by Rekkonnect** [reply]
Example from [here](https://github.com/AlFasGD/GenericsAnalyzer/blob/5b0ea623e606c717e61e86137229d8107756f2a6/GenericsAnalyzer/GenericsAnalyzer.CodeFixes/ConstraintClauseTypeConstraintPlacer.cs#L32):

At L32, I get the semantic model, when I could be getting it much later, since I first use it at L62. With that feature, I could as well do `async await`, and have it prefetched before using it there, while simultaneously taking care of the other operations.

Then, at L101, there are 3 irrelevant `await`s sequentially (L101, L102, L105). They could all be parallelized, since from what I believe neither of those functions interfere with each other (unless I'm understanding the API wrong). The cancellation token would simultaneously notify all tasks anyway, so this is not a problem either.

This might be too specific an example, as it has to do with a code fix that's available on an analyzer, though I remember some other places where I could get a use for that feature.

**Reply by jnm2** [reply]
Leaving aside questions about whether or not parallelizing async operations is actually a deoptimization in the average case, what's the difference between your proposal and the current approach in C#? The L32 example:

```cs
var semanticModel = document.GetSemanticModelAsync(cancellationToken);=
// ...
newConstraintClause = oldConstraintClause.AddUpdateTypeConstraint(
    await semanticModel,
    typeConstraint,
    semanticModel.GetTypeInfo(attributeExpressionType).Type);
```

Is the main difference that you don't have to choose between writing `(await semanticModel)` multiple times or calling the variable `semanticModelTask` and doing `var semanticModel = semanticModelTask`?

There is a bad error handling problem with this code pattern too. Does your proposal have a solution for them? If not, I think it's not a good thing to enshrine in the language. The problem is that if the task fails and the intervening code also happens to throw an exception, the task exception goes unobserved for an arbitrarily long period of time (until next GC) and then it goes to the global exception handler, indistinguishable from any other uncaught exception, and having lost a lot of context that would be useful to report. This unobserved task escalates even if the caller of your method is inside a `try...catch` to catch all exceptions your method throws.

**Reply by jnm2** [reply]
> Then, at L101, there are 3 irrelevant awaits sequentially (L101, L102, L105). They could all be parallelized, since from what I believe neither of those functions interfere with each other (unless I'm understanding the API wrong). The cancellation token would simultaneously notify all tasks anyway, so this is not a problem either.

The second two awaits depend on the result of the first await in your code because you reassign the variable `document`. I believe these will interfere with each other (meaning duplicating cached internal work or one blocks and waits for the other), but nonetheless, you can await the second two parallelizable tasks in parallel easily:

```cs
(semanticModel, root) = await (
    document.GetSemanticModelAsync(cancellationToken),
    document.GetSyntaxRootAsync(cancellationToken));

**Reply by Rekkonnect** [reply]
I don't get why it actually is a deoptimization to parallelize tasks instead of sequentially execute them, in fact, when the async execution overhead is far less than the main operation itself, it should help.

The difference is that indeed I reuse the variable `semanticModel` and update it accordingly once changes apply. Resorting to using `await` to get its result is not my favorite, as it could be reused after having been awaited once. Furthermore, the new solution only adds two more keywords and the variable identifiers in a separate line, helping with code clarity. Which code clarity is lost when using tuple syntax in multiple lines. Additionally, order matters and assignments of resulting values to respective variables is anything but direct. What I mean is that between the variable and its assigned value, there are other variables and expressions whose values are assigned. Such pattern should not be used for larger than 2 to 3 element tuples, I repeat.

I specifically mentioned the 3 lines where there are no interferences, L100 replaces document with the result. That way, with TaskTupleAwaiter, these lines would become:

```csharp
int difference;
(semanticModel, root, difference) = await (
    document.GetSemanticModelAsync(cancellationToken),
    document.GetSyntaxRootAsync(cancellationToken),
    document.LengthDifferenceFrom(oldDocument, cancellationToken));
```

Ignoring the fact that `difference` will have to be separately defined (until C# 10 drops), this multiline tuple requires that you parse the ordering of the variables the values are assigned to. It is not as direct and clean as the feature I proposed allows, which enables clear indication of the assignment.

**Comment by CyrusNajmabadi** [comment]
I genuinely don't see how:

```c#
var result = async await SomeTask();
// other operations
await get result;
```

Is more readable or easier to understand than just:

```c#
var task = SomeTask();
// other operations
var result = await task;
```

It genuinely looks less understandable TBH.

**Reply by Rekkonnect** [reply]
It is the same level of understandable, assuming that you know both concepts. The one just does not introduce a brand new variable to the mix, essentially cleaning out code further. Plus, you won't have to wonder what your task variable will return once executed, if not named `<result>Task`.

**Comment by CyrusNajmabadi** [comment]
> Functions can accept async get parameters, which won't emit the above warning when such a variable is passed, and enabling usage of await get on the passed argument.

Isn't this eqiuvalent to just a method that takes a Task?  A task already represents something whose value can be 'get'ed 'async'hronously.  So having an 'async get' variable just seems like a more verbose way to describe what a Task<> already is.

**Reply by Rekkonnect** [reply]
Yes, but it also enables not using `async get` values. Similarly, if you've already got `async get` variables, you won't have to change your implementation to get the corresponding `Task` to use it in the function.

**Comment by CyrusNajmabadi** [comment]
>  to indicate awaiting the result in a background thread,

I think we would likely be very opposed to that.  'async/await' is about dealing with asynchronous computation of values.  How those values are computed is very out of scope, and very domain specific.  I don't think we'd ever bless something to state that this was computed on a BG thread.  The expectation would be that that is determined by the user themselves.

**Reply by Rekkonnect** [reply]
Updated wording, I did not intend to pass that it enforces a background thread be used.

**Comment by CyrusNajmabadi** [comment]
> but any more will cause worse code quality than it generally would help offering, dragging you back to the starting point.

What examples have more htan this?  I'm trying to figure out the cases where you would independently have so many items in flight, but would then want to join on a large subset of htem.  The proposed syntax of `await get result1, result2;` doesn't seem that much better htan the Tuple-Task approach.

**Comment by yaakov-h** [comment]
Considering the 'recent' work done around async, including ASP.NET Core by default making request/response streams only available through async APIs and not available through synchronous APIs, I suspect there is little to no appetite for a language keyword that blocks execution to wait for the completion of an asynchronous operation.

Particularly not one that only works for Task and not for other awaitable objects that may or may not be trivial to synchronously wait for without incurring further performance costs.

**Reply by Rekkonnect** [reply]
That new feature should be applicable to every case where `await` can normally be used, because it can be backed up by the awaiter that can be returned. I might have to rephrase some things for clarity regarding this aspect.

**Reply by yaakov-h** [reply]
How would that work?

**Reply by Rekkonnect** [reply]
Each variable with `async await` applied to it, begins execution of the respective task in the thread it is assigned to, just like how an async function is normally executed. Then, the awaiter is retrieved, and assigns the result to the variable upon finishing execution, or forces awaiting for execution to be finished with `await get`. Will update the example through using the `GetAwaiter` method.

**Reply by HaloFour** [reply]
@AlFasGD 

That's exactly how all of this works today without your proposed keywords.  `async` doesn't change how a method will execute.

I simply don't see the point of this proposal.  You can already write code that does all of this without any language changes and without additional typing.  Your proposal only makes something that's currently quite simple and makes it muddy.

**Reply by Rekkonnect** [reply]
But the point of the proposal is to add the ability to make parallel execution more flexible while getting rid of temporary variables. It is niche, but I would appreciate its existence in certain scenarios in code that I've actually written.

**Reply by HaloFour** [reply]
I don't see how it makes parallel execution more flexible.  The caller of an async method or expression has no control over that.

As for removing temporary variables, all you're doing is replacing them with a bunch of confusing keywords.  There's nothing wrong with temporary variables and it's not a goal to try to eliminate them from the language.

**Reply by CyrusNajmabadi** [reply]
Ok. @HaloFour 's phrasing helped this click for me. Namely that the goal in your end is to not have both the task variable and the result variable. 

I can somewhat understand that. But I think this approach really feels like a lot of added ceremony for low gain at the end of the day. 

Imo, it's *easier* to understand just using Task and Task-manipulating functions. And it's nice and clear when things transition out. 

I also get much do not like the idea of any sort of implicit blocking. 

--

Note: this feels like an either/or idea here for me. In other words, we might have taken this sort of design when originally doing async/await. However, we went a different route, and I think having both around*at the same time* is actually problematic as it makes this complex space more difficult and confusing, with marginal gain.

**Reply by theunrepentantgeek** [reply]
> while getting rid of temporary variables

Why are temporary variables bad? They give you a chance to give something an intention revealing name, making things clearer for the person who will maintain the code, and in this case they're far to be preferred over some obscure syntax.

I've seen too much code contorted into weird shapes by developers hell-bent on eliminating local variables at the cost of clarity and readability. I think you'd need to go back as far as languages like dBase III and Clipper S87 before you encountered cases where each additional variable had a performance cost ... and _even then_ it was almost always bad idea to to avoid them.
